### PR TITLE
Reapply "VACMS-16853: next build branches ftw (#17154)" (#17261)

### LINKED ISSUE #16853: Next Build and Vets Website branches can be chosen on Tugboat
## Requirements

We need to be able to select different Next Build and Vets Website branches on Tugboat, so that engineers can preview their front-end work in conjunction with CMS changes

```[tasklist]
### Acceptance criteria
- [x] Create a form that allows the user to select Next Build & Vets Website branches on Tugboat
- [x] Upon submission, Next Build will rebuild  for that Tugboat instance based on the branch choices
- [x] The page provides some feedback as to the state of Next Build for that Tugboat environment
```

## Background & implementation details

This will be heavily based on the framework available in vets_website_content_release and vets_website_build_trigger. More concrete information will come.

### PR DESCRIPTION
**Pull request by alexfinnarn**
## Description

closes #16853

## Testing done

- Updated some tests for git information/API/frontend. 
- Manually tested locally by running script.
- Tested on Tugboat

## QA steps

1. Log in and go to "/admin/content/deploy/next"
   - [x] Confirm you can select versions of next-build and vets-website from git references
   - [x] Confirm there is no build in progress
   - [x] No request or lock files should exist
   - [x] The versions for next-build and vets-website should be "__default"
2. Submit the form with two branches chosen for next-build and vets-website
   - [x] Confirm that a build request file exists. 
   - [x] No build should have started yet so no lock file should exist.
   - [x] The versions for next-build and vets-website should be reported as what you selected.
2. Refresh the page a few times until the background process kicks off from reading the build request file.
   - [x] Confirm that the form fields are locked during the build.
   - [x] Confirm you can view the build log file and it is adding data until the build finishes.
4. Refresh the form page after the build is complete
   - [x] Confirm the disabled fields are no longer disabled.
   - [x] View the log to see that the build finished. It might have an error, but that will be addressed in future tickets.
 5. Navigate to "/admin/content/deploy"
    - [x] Confirm that the vets-website version is not the same as the one you chose on "/admin/content/deploy/next".

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.
- [ ] If there are field changes, front end output has been thoroughly checked.

### Select Team for PR review

- [x] `CMS Team`
- [ ] `Public websites`
- [ ] `Facilities`
- [ ] `User support`
- [x] `Accelerated Publishing`

### CODE REVIEW

File: docroot/modules/custom/va_gov_content_release/src/Form/NextGitForm.php [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+        'default' => $this->t('Use default - the vets-website version from the time this demo environment was created.'),
+        'choose' => $this->t('Select a different vets-website branch/pull request - for example, to see your content in a newer frontend design.'),
+      ],
+      '#default_value' => 'default',
+    ];
+
+    $form['build_request']['vets_website_git_ref'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Select branch/pull request'),
+      '#description' => $this->t('Start typing to select a branch for the vets-website version you want to use.'),
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_content_release/src/Form/NextGitForm.php:137]:
```suggestion
      '#description' => $this->t('Start typing to select a branch for the vets-website version you want to use. Note: this vets-website branch is not connected to the vets-website branch used by content-build on this Tugboat instance. If you are testing vets-website changes with both next-build and content-build, both will need the branch updated.'),
```

File: docroot/modules/custom/va_gov_content_release/src/Form/NextGitForm.php [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+   *
+   * @param array $form
+   *   Default form array structure.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   Object containing current form state.
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $form['build_request']['description'] = [
+      '#prefix' => '<br><p>',
+      '#markup' => $this->t('Release content to update the front end of this environment with the latest published content changes.'),
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_content_release/src/Form/NextGitForm.php:93]:
```suggestion
      '#markup' => $this->t('Choose specific branches for the next-build preview server and restart the server. Note: this does not perform any content-build actions.'),
```

**Review by timcosgrove — COMMENTED** [review]
Just putting language changes in here.

File: docroot/modules/custom/va_gov_content_release/va_gov_content_release.routing.yml [hunk 3]
```diff
@@ -16,6 +16,14 @@ va_gov_content_release.form.git:
   requirements:
     _permission: "va gov deploy content build"
 
+va_gov_content_release.form.next_git:
+  path: "/admin/content/deploy/next"
+  defaults:
+    _title: 'Next Build manual content release'
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_content_release/va_gov_content_release.routing.yml:22]:
```suggestion
    _title: 'Choose code branch for Next Build server'
```

File: docroot/modules/custom/va_gov_content_release/src/Form/NextGitForm.php [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+      '#hidden' => TRUE,
+      '#states' => [
+        'visible' => [':input[name="vets_website_selection"]' => ['value' => 'choose']],
+      ],
+    ];
+
+    $form['build_request']['actions']['#type'] = 'actions';
+    $form['build_request']['actions']['submit'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Release Content'),
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_content_release/src/Form/NextGitForm.php:154]: Restart Next Build Server

File: READMES/cms-content-release.md [hunk 5]
```diff
@@ -144,6 +143,35 @@ The Tugboat and local development versions of the release content page do not tr
...
+The upcoming static frontend "next-build" can be rebuilt using different versions of next-build and vets-website. It
+is a simpler process than the current content-build workflow.
+
+1. Go to "/admin/content/deploy/next".
+2. If the form elements are disabled, then a lock file exists preventing another build from being triggered. You
+   can skip to step #6.
+2. Choose a version for next-build or leave at default.
+3. Choose a version for vets-website or leave at default. When content-build is releasing, these form fields might
+   be disabled. We can't change the vets-website version while another frontend build is running.
+4. Click "Release Content" to set the versions of next-build and vets-website as well as write a "buildrequest" file.
```
Comment (timcosgrove) [inline READMES/cms-content-release.md:157]:
```suggestion
4. Click "Restart Next Build Server" to set the versions of next-build and vets-website as well as write a "buildrequest" file.
```

File: READMES/cms-content-release.md [hunk 6]
```diff
@@ -144,6 +143,35 @@ The Tugboat and local development versions of the release content page do not tr
...
+### Next Build Releases
+
+The upcoming static frontend "next-build" can be rebuilt using different versions of next-build and vets-website. It
+is a simpler process than the current content-build workflow.
+
+1. Go to "/admin/content/deploy/next".
+2. If the form elements are disabled, then a lock file exists preventing another build from being triggered. You
+   can skip to step #6.
+2. Choose a version for next-build or leave at default.
+3. Choose a version for vets-website or leave at default. When content-build is releasing, these form fields might
```
Comment (timcosgrove) [inline READMES/cms-content-release.md:155]: Do we know why this locking happens on the next-build form? Can we prevent it?

File: READMES/cms-content-release.md [hunk 7]
```diff
@@ -144,6 +143,35 @@ The Tugboat and local development versions of the release content page do not tr
...
+   be disabled. We can't change the vets-website version while another frontend build is running.
+4. Click "Release Content" to set the versions of next-build and vets-website as well as write a "buildrequest" file.
+5. A `scripts/queue_runner/next_queue_runner.sh` script continuously runs in the background looking for the
+   "buildrequest" file and then start a build if found. Locally, the script has to be triggered manually. See the
+   [caveats](#caveats) section for more information.
+6. Back on "/admin/content/deploy/next" you can view the build log via a link in the "Status" section of the
```
Comment (timcosgrove) [inline READMES/cms-content-release.md:161]:
```suggestion
6. Back on "/admin/content/deploy/next" you can view the server restart log via a link in the "Status" section of the
```

File: READMES/cms-content-release.md [hunk 8]
```diff
@@ -144,6 +143,35 @@ The Tugboat and local development versions of the release content page do not tr
...
+   "Next Build Information" block.
+7. Once the build completes no new build will be triggered until you click to release content again.
```
Comment (timcosgrove) [inline READMES/cms-content-release.md:163]:
```suggestion
7. Once the restart completes no new restart will be triggered unless you request another restart.
```

File: docroot/modules/custom/va_gov_content_release/src/Form/NextGitForm.php (hunk 4 above)
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_content_release/src/Form/NextGitForm.php:154]:
```suggestion
      '#value' => $this->t('Restart Next Build Server'),
```

**Comment by alexfinnarn** [comment]
reopen to kick Tugboat

**Review by tjheffner — APPROVED** [review]
One minor quibble about language in the readme, but QA steps check out, and a different version of vets-website is used for the existing content release form and the new `content/deploy/next` form. 

Two things I notice --
The link to the last build log disappears once the build has completed, I think it would be worthwhile to display a link to the log file if it has been built prior, even if a build is not currently in progress.
Screenshot 2024-02-23 at 10 46 50 AM

Also we get new vets-website branch, but we don't actually use the assets from it because we aren't rebuilding/pulling. I don't think that's a blocker for this ticket, can be addressed as part of #17209.


Overall, great work.

Noting this cannot be merged until the next-build and vets-website-for-next-build repos are added as part of the BRD process, otherwise staging will break again.

File: READMES/cms-content-release.md [hunk 9]
```diff
@@ -144,6 +143,34 @@ The Tugboat and local development versions of the release content page do not tr
 
 For more information on creating or releasing content from a preview environment, see [Environments](./environments.md).
 
+### Next Build Releases
+
+The upcoming static frontend "next-build" can be rebuilt using different versions of next-build and vets-website. It
```
[the remark this answers is not in the document]
Comment (alexfinnarn) [reply READMES/cms-content-release.md:148]: I'm fine with whatever changes would best reflect what is happening when the form gets saved. I know @timcosgrove made updates to the language and I think I kept them all, but maybe I did not.

**Comment by alexfinnarn** [comment]
@tjheffner I added the "display a link to the log file if it has been built prior, even if a build is not currently in progress" note as an AC criterion to https://github.com/department-of-veterans-affairs/va.gov-cms/issues/17209. It shouldn't require much effort.

**Comment by alexfinnarn** [comment]
I am closing this for the time being so Tugboat has more resources, the last build failed anyway, and it can't be merged until https://github.com/department-of-veterans-affairs/devops/pull/14070 is merged.

**Pull request closed without merging**